cache-utils.go

```diff
@@ -0,0 +1,165 @@
+	ch           chan bool   // blocks until header is written or Write started
+	sentResponse bool
+	wroteHeader  bool
+	pr           *io.PipeReader
+	pw           *io.PipeWriter
+}
+
+// NewRecorder returns an initialized ResponseRecorder.
+func NewRecorder() *ResponseRecorder {
+	pr, pw := io.Pipe()
+	return &ResponseRecorder{
+		ch:         make(chan bool),
+		header:     make(http.Header),
+		StatusCode: 200,
+		pw:         pw,
+		pr:         pr,
+	}
+
+}
+
+// Header needed for implementing "net/http".ResponseWriter
+func (rw *ResponseRecorder) Header() http.Header {
+	return rw.header
+}
+
+// writeHeader writes status code if no header written yet.
+func (rw *ResponseRecorder) writeHeader() {
+	if rw.wroteHeader {
+		return
+	}
+	rw.WriteHeader(200)
+}
+
+// Write implements http.ResponseWriter. The data in buf is written to
+// the pipeWriter
+func (rw *ResponseRecorder) Write(buf []byte) (int, error) {
+	rw.writeHeader()
+	if !rw.sentResponse {
+		rw.sendResponse()
+	}
+	return rw.pw.Write(buf)
+}
+
+func (rw *ResponseRecorder) sendResponse() {
+	if rw.sentResponse {
+		return
+	}
+	rw.sentResponse = true
+	rw.ch <- true
+}
+
+// cleanup pipe writer after proxy ServeHTTP finishes.
+func (rw *ResponseRecorder) finish() {
+	if !rw.wroteHeader {
+		rw.WriteHeader(200)
+	}
+	if !rw.sentResponse {
+		rw.sendResponse()
+	}
+	rw.pw.Close()
+	rw.Flush()
+}
+
+// WriteHeader implements http.ResponseWriter.
+func (rw *ResponseRecorder) WriteHeader(code int) {
+	if rw.wroteHeader {
+		return
+	}
+	rw.StatusCode = code
+	rw.wroteHeader = true
+	if rw.header == nil {
+		rw.header = make(http.Header)
+	}
+	if !rw.sentResponse {
+		rw.sendResponse()
+	}
+}
+
+// Flush calls underlying Flush method
+func (rw *ResponseRecorder) Flush() {
+	if !rw.wroteHeader {
+		rw.WriteHeader(200)
+	}
+	rw.Flushed = true
+}
+
+// Result returns the response generated by the handler. It blocks on the
+// rw.ch until header some content has been written
+// The returned Response will have at least its StatusCode,
+// Header, Body
+func (rw *ResponseRecorder) Result() *http.Response {
+	// block until header/writes are started
+	<-rw.ch
+
+	res := &http.Response{
+		Proto:      "HTTP/1.1",
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		StatusCode: rw.StatusCode,
+		Header:     rw.header.Clone(),
+	}
+	res.Body = rw.pr
+	contentLengthStr := res.Header.Get("Content-Length")
+	if contentLengthStr != "" {
+		size, err := strconv.ParseInt(contentLengthStr, 10, 64)
+		if err != nil {
+			res.ContentLength = -1
+		} else {
+			res.ContentLength = size
+		}
+	}
+	return res
+}
+
+type multiWriter struct {
+	backendWriter io.Writer
+	cacheWriter   *io.PipeWriter
+}
+
+func (t *multiWriter) Write(p []byte) (n int, err error) {
+	// for idx, w := range t.writers {
+	n, err = t.backendWriter.Write(p)
+	if err != nil {
+		return
+	}
+	if n != len(p) {
+		err = io.ErrShortWrite
+		return
+	}
+	// ignore errors writing to cache
```

Comment:
close the pipeWriter.CloseWithError(err) with the error returned from Write() call this way the go-routine for PutObject automatically dies. and also you don't have to keep writing after the failure.